File: security-utils.js
/**
 * Security utilities for Node.js/JavaScript applications
 * 
 * Provides common security functions including:
 * - Input validation and sanitization
 * - Token generation
 * - Password hashing (requires bcrypt package)
 * - XSS protection
 */

const crypto = require('crypto');

class SecurityUtils {
  /**
   * Generate a cryptographically secure random token
   * @param {number} length - Length in bytes (default 32)
   * @returns {string} Hex string token
   */
  static generateSecureToken(length = 32) {
    return crypto.randomBytes(length).toString('hex');
  }

  /**
   * Sanitize user input to prevent XSS attacks
   * @param {string} input - User input to sanitize
   * @param {number} maxLength - Maximum allowed length
   * @returns {string} Sanitized string
   */
  static sanitizeInput(input, maxLength = 1000) {
    if (!input || typeof input !== 'string') {
      return '';
    }

    // Truncate to max length
    let sanitized = input.slice(0, maxLength);

    // Escape HTML special characters
    const htmlEscapeMap = {
      '&': '&amp;',
      '<': '&lt;',
      '>': '&gt;',
      '"': '&quot;',
      "'": '&#x27;',
      '/': '&#x2F;',
    };

    sanitized = sanitized.replace(/[&<>"'\/]/g, (char) => htmlEscapeMap[char]);

    return sanitized.trim();
  }

  /**
   * Validate email format
   * @param {string} email - Email to validate
   * @returns {boolean} True if valid format
   */
  static validateEmail(email) {
    const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
    return emailRegex.test(email);
  }

  /**
   * Validate URL format
   * @param {string} url - URL to validate
   * @returns {boolean} True if valid format
   */
  static validateURL(url) {
    try {
      const urlObj = new URL(url);
      return ['http:', 'https:'].includes(urlObj.protocol);
    } catch {
      return false;
    }
  }

  /**
   * Generate a secure hash of data
   * @param {string} data - Data to hash
   * @param {string} algorithm - Hash algorithm (default: sha256)
   * @returns {string} Hex string hash
   */
  static hash(data, algorithm = 'sha256') {
    return crypto.createHash(algorithm).update(data).digest('hex');
  }

  /**
   * Create HMAC for data integrity verification
   * @param {string} data - Data to sign
   * @param {string} secret - Secret key
   * @returns {string} HMAC hex string
   */
  static createHMAC(data, secret) {
    return crypto.createHmac('sha256', secret).update(data).digest('hex');
  }

  /**
   * Verify HMAC signature
   * @param {string} data - Original data
   * @param {string} secret - Secret key
   * @param {string} signature - HMAC to verify
   * @returns {boolean} True if signature matches
   */
  static verifyHMAC(data, secret, signature) {
    const expectedSignature = this.createHMAC(data, secret);
    return crypto.timingSafeEqual(
      Buffer.from(signature),
      Buffer.from(expectedSignature)
    );
  }

  /**
   * Rate limiting helper - simple in-memory implementation
   * WARNING: Use Redis or similar in production for distributed systems
   */
  static createRateLimiter() {
    const requests = new Map();
    
    return {
      /**
       * Check if request should be allowed
       * @param {string} identifier - Unique identifier (IP, user ID, etc.)
       * @param {number} maxRequests - Max requests allowed
       * @param {number} windowMs - Time window in milliseconds
       * @returns {boolean} True if request allowed
       */
      isAllowed(identifier, maxRequests = 100, windowMs = 60000) {
        const now = Date.now();
        const userRequests = requests.get(identifier) || [];
        
        // Remove old requests outside the window
        const validRequests = userRequests.filter(
          (timestamp) => now - timestamp < windowMs
        );
        
        if (validRequests.length >= maxRequests) {
          return false;
        }
        
        validRequests.push(now);
        requests.set(identifier, validRequests);
        return true;
      },
      
      /**
       * Clear rate limit data for identifier
       * @param {string} identifier - Identifier to clear
       */
      clear(identifier) {
        requests.delete(identifier);
      }
    };
  }
}

// Example usage
if (require.main === module) {
  console.log('Security Utils Examples:');
  
  // Generate token
  const token = SecurityUtils.generateSecureToken();
  console.log('Secure token:', token);
  
  // Sanitize input
  const userInput = '<script>alert("xss")</script>Hello';
  const sanitized = SecurityUtils.sanitizeInput(userInput);
  console.log('Sanitized:', sanitized);
  
  // Validate email
  const email = 'user@example.com';
  console.log('Email valid:', SecurityUtils.validateEmail(email));
  
  // Hash data
  const hash = SecurityUtils.hash('sensitive data');
  console.log('Hash:', hash);
  
  // HMAC example
  const data = 'important message';
  const secret = 'secret-key';
  const hmac = SecurityUtils.createHMAC(data, secret);
  console.log('HMAC:', hmac);
  console.log('HMAC valid:', SecurityUtils.verifyHMAC(data, secret, hmac));
}

module.exports = SecurityUtils;

File: security_utils.py
"""
Security utilities and helper functions for the application.

This module provides security-related functions including:
- Password hashing and verification
- Token generation and validation
- Input sanitization
- Secure random generation
"""

import secrets
import hashlib
import re
from typing import Optional


class SecurityUtils:
    """Security utility class with common security functions."""

    @staticmethod
    def generate_secure_token(length: int = 32) -> str:
        """
        Generate a cryptographically secure random token.

        Args:
            length: Length of the token in bytes (default 32)

        Returns:
            Hex string of the generated token
        """
        return secrets.token_hex(length)

    @staticmethod
    def hash_password(password: str) -> str:
        """
        Hash a password using SHA-256 with salt.
        
        WARNING: This is a basic example. In production, use:
        - bcrypt (recommended)
        - argon2 (recommended)
        - PBKDF2
        
        Args:
            password: Plain text password
            
        Returns:
            Hashed password with salt
        """
        salt = secrets.token_hex(16)
        pwd_hash = hashlib.pbkdf2_hmac('sha256', 
                                        password.encode('utf-8'), 
                                        salt.encode('utf-8'), 
                                        100000)
        return f"{salt}${pwd_hash.hex()}"

    @staticmethod
    def verify_password(password: str, hashed: str) -> bool:
        """
        Verify a password against a hash.
        
        Args:
            password: Plain text password to verify
            hashed: Stored password hash
            
        Returns:
            True if password matches, False otherwise
        """
        try:
            salt, pwd_hash = hashed.split('$')
            new_hash = hashlib.pbkdf2_hmac('sha256',
                                           password.encode('utf-8'),
                                           salt.encode('utf-8'),
                                           100000)
            return new_hash.hex() == pwd_hash
        except (ValueError, AttributeError):
            return False

    @staticmethod
    def sanitize_input(user_input: str, max_length: int = 1000) -> str:
        """
        Sanitize user input to prevent injection attacks.
        
        Args:
            user_input: Raw user input
            max_length: Maximum allowed length
            
        Returns:
            Sanitized string
        """
        if not user_input:
            return ""
        
        # Truncate to max length
        sanitized = user_input[:max_length]
        
        # Remove potentially dangerous characters
        # Customize based on your specific needs
        sanitized = re.sub(r'[<>\"\'&;]', '', sanitized)
        
        return sanitized.strip()

    @staticmethod
    def validate_email(email: str) -> bool:
        """
        Validate email format.
        
        Args:
            email: Email address to validate
            
        Returns:
            True if valid format, False otherwise
        """
        pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        return bool(re.match(pattern, email))


# Example usage (for documentation purposes)
if __name__ == "__main__":
    utils = SecurityUtils()
    
    # Generate secure token
    token = utils.generate_secure_token()
    print(f"Secure token: {token}")
    
    # Password hashing example
    password = "example_password"
    hashed = utils.hash_password(password)
    print(f"Hashed password: {hashed}")
    
    # Verify password
    is_valid = utils.verify_password(password, hashed)
    print(f"Password valid: {is_valid}")
    
    # Sanitize input
    user_input = "<script>alert('xss')</script>Hello"
    clean = utils.sanitize_input(user_input)
    print(f"Sanitized: {clean}")
    
    # Validate email
    email = "user@example.com"
    is_valid_email = utils.validate_email(email)
    print(f"Email valid: {is_valid_email}")
